src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Runtime
+{
+    /// <summary>
+    /// Allows to run code and abort it asynchronously.
+    /// </summary>
+    public static partial class ControlledExecution
+    {
+        /// <summary>
+        /// Runs code that may be aborted asynchronously.
+        /// </summary>
+        /// <param name="action">The delegate that represents the code to execute.</param>
+        /// <param name="cancellationToken">The cancellation token that may be used to abort execution.</param>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="action"/> argument is null.</exception>
+        /// <exception cref="System.InvalidOperationException">
+        /// The current thread is already running the <see cref="ControlledExecution.Run"/> method.
+        /// </exception>
+        /// <exception cref="System.OperationCanceledException">The execution was aborted.</exception>
+        [Obsolete(Obsoletions.ControlledExecutionRunMessage, DiagnosticId = Obsoletions.ControlledExecutionRunDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
+        public static void Run(Action action, CancellationToken cancellationToken)
+        {
+            ArgumentNullException.ThrowIfNull(action);
+            var execution = new Execution(action, cancellationToken);
+            cancellationToken.Register(execution.Abort, useSynchronizationContext: false);
+            execution.Run();
```

Comment:
Could you please explain the difference regarding `UnsafeRegister`?  Do we have to use a lambda here?

Reply:
The only difference between Register and UnsafeRegister is whether ExecutionContext is captured and used to invoke the delegate: Register does, UnsafeRegister doesn't.  And you only need the ExecutionContext if the code to be invoked is going to execute user code; ExecutionContext flows AsyncLocals.  Since the callback doesn't invoke any user code, it can use UnsafeRegister.

As for lambdas, no, the method takes a delegate.  You can pass a method group and have the C# compiler create a delegate for you from that method group, or you can pass a lambda and have the C# compiler create a delegate for you from that lambda.  The overload you were calling, though, takes an Action, and every call with `execution.Abort` is going to allocate a new Action.  The overload I used takes an `Action<object>` and will feed in the `object state` argument passed as the second argument to {Unsafe}Register.  With that, the lambda doesn't close over any instance state, and thus the C# compiler caches a delegate that it will reuse on every call, such that you only get one delegate allocated for all calls vs one delegate allocated per call.

Reply:
> And you only need the ExecutionContext if the code to be invoked is going to execute user code; ExecutionContext flows AsyncLocals. Since the callback doesn't invoke any user code, it can use UnsafeRegister.

What if a user registers their own callback on this `_cancellationToken`?

Reply:
Ah, never mind, they would use a normal `Register`.